### PR TITLE
[SPARK-2808][Streaming][Kafka] update kafka to 0.8.2

### PR DESCRIPTION
**Pull request by koeninger**
i don't think this should be merged until after 1.3.0 is final

### CODE REVIEW

**Comment by srowen** [comment]
ok to test

**Comment by zzcclp** [comment]
Will this RP be merged into 1.3.0?

**Comment by koeninger** [comment]
I believe that it will not be merged into 1.3

**Comment by luisobo** [comment]
@koeninger what's the reason for not putting this into 1.3.0? Kafka 0.8.2 adds support for scala 2.11. That plus [non-experimental scala 2.11 support in spark](https://issues.apache.org/jira/browse/SPARK-5850) is an interesting combo for scala 2.11 projects. 

Thanks in any case.

**Comment by koeninger** [comment]
Spark 1.3.0 was already supposed to be frozen a while back as far as I know.

My personal gut feeling is that it'd be better to wait for kafka 0.8.2.1
bugfixes and more of a chance to test.

**Comment by zzcclp** [comment]
@koeninger , kafka 0.8.2.1 was [released](https://archive.apache.org/dist/kafka/0.8.2.1/RELEASE_NOTES.html), can it be upgrade this version?

**Comment by koeninger** [comment]
Those tests have passed locally for me 3 times in a row... if I get time later I'll try to dig in

**Comment by zeitos** [comment]
+1 I just tested it with 2.10 and 2.11 and it also passed, can an admin retest?

**Comment by srowen** [comment]
Jenkins, retest this please

**Comment by zzcclp** [comment]
@koeninger , no progress yet?

**Comment by koeninger** [comment]
@zzcclp if you want to help try and figure out how reproduce this test failure outside of Jenkins, go for it.

**Comment by zzcclp** [comment]
@koeninger , I can't visit [this url](https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28872/) , it's 404. ??

**Comment by srowen** [comment]
Jenkins, retest this please

**Comment by zzcclp** [comment]
there is not any error, why?

**Comment by zzcclp** [comment]
@koeninger , I merge this RP into master and build, it is OK. then run tests as follows:
>mvn -Pscala-2.10 -Phadoop-2.3 -Pyarn -Dyarn.version=2.3.0-cdh5.1.2 -Dhadoop.version=2.3.0-cdh5.1.2 -pl core,streaming,external/kafka -DwildcardSuites=org.apache.spark.streaming.kafka.KafkaClusterSuite test

>mvn -Pscala-2.10 -Phadoop-2.3 -Pyarn -Dyarn.version=2.3.0-cdh5.1.2 -Dhadoop.version=2.3.0-cdh5.1.2 -pl core,streaming,external/kafka -DwildcardSuites=org.apache.spark.streaming.kafka.KafkaStreamSuite test

all are successful.

I am new to Spark and only do little help.

**Comment by koeninger** [comment]
As far as I can tell from jenkins output, it failed during MiMa checks.

But if I run

 sbt '; project streaming-kafka; mima-report-binary-issues'

I get

[info] spark-streaming-kafka: found 0 potential binary incompatibilities (filtered 1)

**Comment by srowen** [comment]
Jenkins claims ...

```
[error]  * method setConsumerOffsetMetadata(java.lang.String,scala.collection.immutable.Map)scala.util.Either in class org.apache.spark.streaming.kafka.KafkaCluster does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.streaming.kafka.KafkaCluster.setConsumerOffsetMetadata")
[error]  * method setConsumerOffsets(java.lang.String,scala.collection.immutable.Map)scala.util.Either in class org.apache.spark.streaming.kafka.KafkaCluster does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.streaming.kafka.KafkaCluster.setConsumerOffsets")
[error]  * method getConsumerOffsetMetadata(java.lang.String,scala.collection.immutable.Set)scala.util.Either in class org.apache.spark.streaming.kafka.KafkaCluster does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.streaming.kafka.KafkaCluster.getConsumerOffsetMetadata")
[error]  * method getConsumerOffsets(java.lang.String,scala.collection.immutable.Set)scala.util.Either in class org.apache.spark.streaming.kafka.KafkaCluster does not have a correspondent in new version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.streaming.kafka.KafkaCluster.getConsumerOffsets")
```

**Comment by koeninger** [comment]
That maybe makes a certain amount of sense... I'll try replacing the default arguments with multiple overloaded methods, see if that passes mima.

**Comment by zzcclp** [comment]
@koeninger , I test streaming + kakfa 0.8.2.1 with this RP, it works well, :smile:

**Comment by koeninger** [comment]
Glad to hear it works for you. Fixing the default arguments for mima was straightforward, but there's a lot that has changed in the test code.  Work's been busy, haven't had time to fix the merge conflicts yet

**Comment by zzcclp** [comment]
@koeninger , will this RP be merged into 1.4.0?

**Comment by koeninger** [comment]
@zzcclp the 1.4.0 deadline is Friday, I believe.  I fixed the merge conflicts and resolved the MiMa issue (as far as I know), it still passes local tests for me.  We'll see if it passes on jenkins once an admin verifies it.  It sounds like there was a large build backlog due to jenkins server issues over the weekend.

**Comment by luisobo** [comment]
Please, let me know if there is anything I could do to help get this in
1.4.0. I really need this one.

Thank you!

2015-04-28 10:34 GMT-04:00 Cody Koeninger <notifications@github.com>:

**Comment by koeninger** [comment]
@srowen I imagine things are pretty busy, but can you verify this to test on jenkins?
@luisobo you could build and test it out

**Comment by srowen** [comment]
ok to test

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala [hunk 1]
```diff
@@ -220,12 +220,22 @@ class KafkaCluster(val kafkaParams: Map[String, String]) extends Serializable {
...
+  def defaultConsumerApiVersion: Short = 0
+
   /** Requires Kafka >= 0.8.1.1 */
   def getConsumerOffsets(
       groupId: String,
       topicAndPartitions: Set[TopicAndPartition]
+    ): Either[Err, Map[TopicAndPartition, Long]] =
+    getConsumerOffsets(groupId, topicAndPartitions, defaultConsumerApiVersion)
+
+  def getConsumerOffsets(
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: Nit: These two can be merged with default arguments.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala [hunk 2]
```diff
@@ -220,12 +220,22 @@ class KafkaCluster(val kafkaParams: Map[String, String]) extends Serializable {
...
+      groupId: String,
+      topicAndPartitions: Set[TopicAndPartition],
+      versionId: Short
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: Can rename versionId to consumerApiVersion

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala [hunk 3]
```diff
@@ -236,9 +246,16 @@ class KafkaCluster(val kafkaParams: Map[String, String]) extends Serializable {
   def getConsumerOffsetMetadata(
       groupId: String,
       topicAndPartitions: Set[TopicAndPartition]
+    ): Either[Err, Map[TopicAndPartition, OffsetMetadataAndError]] =
+    getConsumerOffsetMetadata(groupId, topicAndPartitions, defaultConsumerApiVersion)
+
+  def getConsumerOffsetMetadata(
+      groupId: String,
+      topicAndPartitions: Set[TopicAndPartition],
+      versionId: Short
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: Similar comment as above.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala [hunk 4]
```diff
@@ -266,24 +283,39 @@ class KafkaCluster(val kafkaParams: Map[String, String]) extends Serializable {
   def setConsumerOffsets(
       groupId: String,
       offsets: Map[TopicAndPartition, Long]
+    ): Either[Err, Map[TopicAndPartition, Short]] =
+    setConsumerOffsets(groupId, offsets, defaultConsumerApiVersion)
+
+  def setConsumerOffsets(
+      groupId: String,
+      offsets: Map[TopicAndPartition, Long],
+      versionId: Short
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: Same comment as above. versionId and default arguments.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala [hunk 5]
```diff
@@ -227,10 +229,33 @@ private class KafkaTestUtils extends Logging {
     tryAgain(1)
   }
 
+  def waitUntilLeaderOffset(
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala]: What does this function do? Either make the name more meaningful (like `waitUntilMetadataIsPropagated`) or add scala docs.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala [hunk 6]
```diff
@@ -53,14 +53,16 @@ class KafkaRDDSuite extends FunSuite with BeforeAndAfterAll {
   }
 
   test("basic usage") {
-    val topic = "topicbasic"
+    val topic = s"topicbasic-${Random.nextInt}"
```
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala]: Why was this change necessary?

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala [hunk 7]
```diff
@@ -53,14 +53,16 @@ class KafkaRDDSuite extends FunSuite with BeforeAndAfterAll {
...
     val messages = Set("the", "quick", "brown", "fox")
     kafkaTestUtils.sendMessages(topic, messages.toArray)
 
-
     val kafkaParams = Map("metadata.broker.list" -> kafkaTestUtils.brokerAddress,
-      "group.id" -> s"test-consumer-${Random.nextInt(10000)}")
+      "group.id" -> s"test-consumer-${Random.nextInt}")
+
+    val kc = new KafkaCluster(kafkaParams)
+    kafkaTestUtils.waitUntilLeaderOffset(kc, topic, 0, messages.size)
```
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala]: What was this, and why was this necessary?

**Comment by tdas** [comment]
Now that the Kafka Direct Python API has been merged, I think its safe to pull this in. Overall it looks good. I had a few minor comments on the code. However I do have a high level question. Will this operate with 0.8.1.x Kafka installations?

**Comment by tdas** [comment]
Also it would be great if we can merge this by this weekend. It seems mostly ready thanks to your continuous rebasing.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala (hunk 1 above)
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: The code was originally using default arguments.  That's what was causing the MiMa binary compatibility errors

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala (hunk 7 above)
Comment (koeninger) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala]: Those were just ruling out the possibility of jenkins failures being related to stale duplicate topic or consumer state.  There's nothing magical about the number 10000 as a random range for the name, so I removed it.

**Comment by koeninger** [comment]
I can do some basic testing against a non-embedded 0.8.1 kafka install.

We're not likely to upgrade for our production jobs, so if some of the
people who really want this version bumped up could do some testing with
production-like usage, that would be great.

**Comment by koeninger** [comment]
At this point, it looks like the java and scala tests are passing, but the python tests are timing out.  Console output looks like

=========================================================================
Running PySpark tests
=========================================================================
Running PySpark tests. Output is in python/unit-tests.log.
Testing with Python version:
Python 2.6.6
Run core tests ...
(omitted)
Run streaming tests ...
Running test: pyspark/streaming/util.py ... ok (0s)

then it just hangs, apparently at some point during pyspark/streaming/tests.py

@davies I'm not familiar with the python code, do you have any ideas?

**Comment by tdas** [comment]
Is it timing out locally as well?

**Comment by tdas** [comment]
retest this please.

**Comment by koeninger** [comment]
No, it works fine locally:

Running PySpark tests. Output is in python/unit-tests.log.

Testing with Python version:

Python 2.6.6

Run core tests ...

Running test: pyspark/rdd.py ... ok (11s)

(omitted)

Run streaming tests ...

Running test: pyspark/streaming/util.py ... ok (0s)

Running test: pyspark/streaming/tests.py ... ok (84s)

Tests passed in 372 seconds

**Comment by tdas** [comment]
Actually this branch has merge conflicts with the master. May be that is the reason? Please update your branch and test again. 

Also I saw python streaming tests handing in another PR of mine last night. I wonder if this is a Jenkins issue. Nonetheless please update your branch.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala (hunk 1 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: that is frigging weird. This is in an internal class and should not be throwing MIMA compatibility errors. :/

**Comment by tdas** [comment]
That's a whole lot of failures. I am not sure what happened. Running the test once again.

**Comment by tdas** [comment]
Ignore this last failure, this started before the merging. Lets focus on Test Build #734

**Comment by tdas** [comment]
retest this please

**Comment by tdas** [comment]
Hey @koeninger , here is a trick I am trying out to to do targeted debugging in Jenkins. I disabled all the other tests and running only streaming tests. The whole thing runs in less than 20 minutes. 

https://github.com/apache/spark/pull/5824

Try this on your PR to see whether you can pin point the issue?

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala (hunk 7 above)
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala]: That makes sense for the topic name change. This question was about the `waitUntilLeaderOffset`. Is it also to make sure that the leader shows the right expected offset before testing starts? For more reliable testing?

File: python/pyspark/streaming/tests.py [hunk 8]
```diff
@@ -616,6 +615,7 @@ def test_kafka_stream(self):
 
         self._kafkaTestUtils.createTopic(topic)
         self._kafkaTestUtils.sendMessages(topic, sendData)
+        time.sleep(5)
```
Comment (tdas) [inline python/pyspark/streaming/tests.py]: If the `waitUntilLeaderOffset` has been added to Scala, shouldnt they be added to Python as well? Maybe that then this arbitrary sleep wont be needed.

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala (hunk 7 above)
Comment (koeninger) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala]: Yeah, thats' what got the tests passing on Jenkins

**Comment by koeninger** [comment]
scalastyle passes locally for me...

**Comment by koeninger** [comment]
At any rate, I tested it out locally against 0.8.1.1 server install, with the IdempotentExample job from my blogpost.  Seemed to work ok.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala [hunk 9]
```diff
@@ -227,10 +229,34 @@ private class KafkaTestUtils extends Logging {
...
+        llo == offset,
+        s"$topic $partition $offset not reached after timeout")
+    }
+  }
+
   private def waitUntilMetadataIsPropagated(topic: String, partition: Int): Unit = {
     eventually(Time(10000), Time(100)) {
       assert(
-        server.apis.metadataCache.containsTopicAndPartition(topic, partition),
+        server.apis.metadataCache.getPartitionInfo(topic, partition) match {
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala]: Could you put this big piece of code in a separate `def` (can be inside `waitUntilMetadata..`) outside the eventually, and then the assert just calls that function? Much easier to read.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala [hunk 10]
```diff
@@ -37,6 +38,11 @@ private[spark]
 class KafkaCluster(val kafkaParams: Map[String, String]) extends Serializable {
   import KafkaCluster.{Err, LeaderOffset, SimpleConsumerConfig}
 
+  /** Constructor that takes a Java map */
+  def this(kafkaParams: java.util.Map[String, String]) {
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: Is this used anywhere? If it is not, then why was this added?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala [hunk 11]
```diff
@@ -220,12 +226,22 @@ class KafkaCluster(val kafkaParams: Map[String, String]) extends Serializable {
   // https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-OffsetCommit/FetchAPI
   // scalastyle:on
 
+  // this 0 here indicates api version, in this case the original ZK backed api.
+  def defaultConsumerApiVersion: Short = 0
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala]: Shouldn't this be private? Do other parts of the spark/streaming call this function?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala [hunk 12]
```diff
@@ -227,10 +229,34 @@ private class KafkaTestUtils extends Logging {
     tryAgain(1)
   }
 
+  /** wait until the leader offset for the given topic/partition equals the specified offset */
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala]: Nit: wait --> Wait

File: python/pyspark/streaming/tests.py [hunk 13]
```diff
@@ -659,7 +659,7 @@ def test_kafka_rdd(self):
 
         self._kafkaTestUtils.createTopic(topic)
         self._kafkaTestUtils.sendMessages(topic, sendData)
-
+        self._kafkaTestUtils.waitUntilLeaderOffset(topic, 0, sum(sendData.values()))
```
Comment (tdas) [inline python/pyspark/streaming/tests.py]: Should be added for the all the kafka unit tests, whereever there is a sendMessage. Its missing in `test_kafka_direct_stream***`

File: external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala [hunk 14]
```diff
@@ -73,27 +74,37 @@ class KafkaRDDSuite extends FunSuite with BeforeAndAfterAll {
...
+    val ranges = rdd.get.asInstanceOf[HasOffsetRanges].offsetRanges
+    val rangeCount = ranges.map(o => o.untilOffset - o.fromOffset).sum
 
-    kc.setConsumerOffsets(kafkaParams("group.id"), ranges)
+    assert(rangeCount === sentCount, "offset range didn't include all sent messages")
+    assert(rdd.get.count === sentCount, "didn't get all sent messages")
+
+    val rangesMap = ranges.map(o => TopicAndPartition(o.topic, o.partition) -> o.untilOffset).toMap
+
+    kc.setConsumerOffsets(kafkaParams("group.id"), rangesMap).fold(
```
Comment (tdas) [inline external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaRDDSuite.scala]: What is this for? Its not obvious, could you add a comment?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala (hunk 9 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaTestUtils.scala]: And also explain in a comment what this code does? For example, I am not sure what is `isr` in `leaderAndIsr`. And what are all these conditions &&-ed together.

**Comment by tdas** [comment]
Running this couple of times to test for flakiness.

**Comment by tdas** [comment]
Great! Merging this! Thanks @koeninger @helena !
